interfaces/prompting/prompting.go

```diff
@@ -0,0 +1,154 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package prompting
+
+import (
+	"encoding/base32"
+	"encoding/binary"
+	"fmt"
+	"time"
+)
+
+type OutcomeType string
+
+const (
+	OutcomeUnset OutcomeType = ""
+	OutcomeAllow OutcomeType = "allow"
+	OutcomeDeny  OutcomeType = "deny"
+)
+
+// AsBool returns the outcome as a boolean, or an error if it cannot be parsed.
+func (outcome OutcomeType) AsBool() (bool, error) {
+	switch outcome {
+	case OutcomeAllow:
+		return true, nil
+	case OutcomeDeny:
+		return false, nil
+	default:
+		return false, fmt.Errorf(`invalid outcome: must be %q or %q: %q`, OutcomeAllow, OutcomeDeny, outcome)
+	}
+}
+
+// ValidateOutcome returns nil if the given outcome is valid, otherwise an error.
+func ValidateOutcome(outcome OutcomeType) error {
```

Comment:
this is a bit strange, shouldn't it be a parse function taking a string and returning the OutcomeType or an error?

Reply:
or maybe what you need is MarshalJSON UnmarshalJSON methods?

Reply:
Since `OutcomeType` is an alias for `string`, the json {,un}marshaller works on `OutcomeType` directly, so potentially malformed outcomes are already `OutcomeType` variables which need to be validated, rather than strings. Theoretically, this could be built into custom `{M,Unm}arshalJSON` functions, but then there's no way to validate values when explicitly defined manually (i.e. `myOutcome := OutcomeType("allow")` instead of `OutcomeAllow`). There should probably never be a case where custom strings are cast to `OutcomeType`s, so perhaps the ergonomics of a custom `{M,Unm}arshalJSON` would be helpful.

Reply:
Hm if we don't care about defining new values of OutcomeType manually, then I think a trivial UnmarshalJSON would be preferred. You do not need MarshalJSON in such case.